app/services/bucket_service.rb

```diff
@@ -1,12 +1,61 @@
 class BucketService
-  def self.archive(bucket:, exclude_author_from_email_notifications: false)
-    bucket.update(archived_at: DateTime.now.utc)
-    if bucket.status == 'live' || bucket.status == 'funded'
-      bucket.contributors(exclude_author: exclude_author_from_email_notifications).each do |funder|
-        UserMailer.notify_funder_that_bucket_was_archived(funder: funder, bucket: bucket).deliver_later
+  def self.archive(bucket, user, exclude_author_from_email_notifications: false)
+    ActiveRecord::Base.transaction do
+      if bucket.is_idea? || bucket.is_funding? || bucket.is_funded?
+        bucket.contributors(exclude_author: exclude_author_from_email_notifications).each do |funder|
+          # Don't notify archived members
+          if !Membership.find_by(group_id: bucket.group_id, member_id: funder.id).archived_at
+            UserMailer.notify_funder_that_bucket_was_archived(funder: funder, bucket: bucket).deliver_later
+          end
+        end
+        bucket.update(archived_at: DateTime.now.utc)
+        bucket.contributions.destroy_all
```

Comment:
have you thought about not destroying the contributions and instead archiving them? do you think this history interesting and/or useful to preserve?

Reply:
Yes, I think it's essential we preserve the history. This is one of the main changes behind this whole thing. As you can see, the transaction records are not being destroyed. Instead new "reverse" transactions are entered, so we never loose history.

The idea is to eventually get rid of the `contributions` table and the `allocations` table completely so both are completely replaced by the information in the `transactions` table. This should already be the case, as all updates should be reflected in the `transactions`. You will also notice I've created a job to be run daily to check that the old `contributions` and `allocations` system is in alignment with the `tranactions` system.

In reality what I'm trying to do here is to release a change in steps - before it's fully finished. This step is updating all the functions that write money information to update both "systems". Next step will be to change the different functions that read and return data to use the new `transactions` system as well.

Does this make sense?

Reply:
ah yes of course that makes sense! thx for the clarification